src/error.rs

```diff
@@ -0,0 +1,136 @@
+use std::time::Duration;
+use backoff::backoff::Backoff;
+use once_cell::sync::Lazy;
+use crate::{Config, ConfigMeta, clients};
+
+static CLIENT_ERR_REGEX: Lazy<regex::Regex> = Lazy::new(|| regex::Regex::new(r"Client \{ status: (?<status>\d+?),").unwrap());
+static REQWEST_ERR_REGEX: Lazy<regex::Regex> = Lazy::new(|| regex::Regex::new(r"Reqwest \{ retries: (?<retries>\d+?), max_retries: (?<max_retries>\d+?),").unwrap());
+
+#[derive(Debug)]
+pub(crate) struct ErrorInfo {
+    pub(crate) retries: Option<usize>,
+    pub(crate) reason: ErrorReason
+}
+
+#[derive(Debug)]
+pub(crate) enum ErrorReason {
+    Unknown,
+    Code(u16),
+    Io,
+    Timeout
+}
+
+fn backoff_duration_for_retry(retries: usize, meta: &ConfigMeta) -> Duration {
+    let mut backoff = backoff::ExponentialBackoff {
+        initial_interval: meta.retry_config.backoff.init_backoff,
+        max_interval: meta.retry_config.backoff.max_backoff,
+        ..Default::default()
+    };
+
+    for _retry in 0..retries {
+        let _ = backoff.next_backoff();
+    }
+
+    backoff.next_backoff().unwrap_or(meta.retry_config.backoff.max_backoff)
+}
+
+pub(crate) fn extract_error_info(error: &anyhow::Error) -> ErrorInfo {
+    let mut retries = None;
+    for e in error.chain() {
+        if let Some(e) = e.downcast_ref::<reqwest::Error>() {
+            if let Some(code) = e.status() {
+                return ErrorInfo {
+                    retries,
+                    reason: ErrorReason::Code(code.into())
+                }
+            }
+            if e.is_timeout() {
+                return ErrorInfo {
+                    retries,
+                    reason: ErrorReason::Timeout
+                }
+            }
+            if e.is_body() || e.is_connect() || e.is_request() {
+                return ErrorInfo {
+                    retries,
+                    reason: ErrorReason::Io
+                }
+            }
+        }
+
+        if let Some(e) = e.downcast_ref::<hyper::Error>() {
```

Comment:
Handle hyper IO error for TCP reset:
`exception: RustyObjectStore.PutException("failed to process put with error: Generic MicrosoftAzure error: Error after 0 retries in 8.467222256s, max_retries:10, retry_timeout:180s, source:error sending request for url (https://raistagingeastusdb.blob.core.windows.net/transient-test-data/relationalai-gha-transient/data/189/2a96e1ef78a400115f6b542556e10989): connection error: Connection reset by peer (os error 104)")`